include/pybind11/warnings.h

```diff
@@ -0,0 +1,70 @@
+/*
+    pybind11/warnings.h: Python warnings wrappers.
+
+    Copyright (c) 2024 Jan Iwaszkiewicz <jiwaszkiewicz6@gmail.com>
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#pragma once
+
+#include "pybind11.h"
+#include "detail/common.h"
+
+PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+
+PYBIND11_NAMESPACE_BEGIN(detail)
+
+inline bool PyWarning_Check(PyObject *obj) {
+    int result = PyObject_IsSubclass(obj, PyExc_Warning);
+    if (result == 1) {
+        return true;
+    }
+    if (result == -1) {
+        raise_from(PyExc_SystemError,
+                   "PyWarning_Check(): internal error of Python C API while "
+                   "checking a subclass of the object!");
+        throw error_already_set();
+    }
+    return false;
+}
+
+PYBIND11_NAMESPACE_END(detail)
+
+PYBIND11_NAMESPACE_BEGIN(warnings)
+
+inline object
+new_warning_type(handle scope, const char *name, handle base = PyExc_RuntimeWarning) {
+    if (!detail::PyWarning_Check(base.ptr())) {
+        pybind11_fail("warning(): cannot create custom warning, base must be a subclass of "
+                      "PyExc_Warning!");
+    }
+    if (hasattr(scope, "__dict__") && scope.attr("__dict__").contains(name)) {
+        pybind11_fail("Error during initialization: multiple incompatible "
```

Comment:
Maybe
```
pybind11_fail("new_warning_type(): an attribute with name \"" + std::string(name) + "\" exists already.");
```
?

I'm also wondering, could the condition simply be `hasattr(scope, name)`? (I'm really not sure.)

Reply:
You are right, it works the same. Also added test case for it.

For the message same as https://github.com/pybind/pybind11/pull/5291#discussion_r1717763599